src/base/bittorrent/dbresumedatastorage.cpp

```diff
@@ -604,216 +627,288 @@ void BitTorrent::DBResumeDataStorage::enableWALMode() const
         throw RuntimeError(tr("WAL mode is probably unsupported due to filesystem limitations."));
 }
 
-BitTorrent::DBResumeDataStorage::Worker::Worker(const Path &dbPath, const QString &dbConnectionName, QReadWriteLock &dbLock)
+BitTorrent::DBResumeDataStorage::Worker::Worker(const Path &dbPath, QReadWriteLock &dbLock)
     : m_path {dbPath}
-    , m_connectionName {dbConnectionName}
     , m_dbLock {dbLock}
 {
 }
 
-void BitTorrent::DBResumeDataStorage::Worker::openDatabase() const
+void BitTorrent::DBResumeDataStorage::Worker::run()
 {
-    auto db = QSqlDatabase::addDatabase(u"QSQLITE"_qs, m_connectionName);
-    db.setDatabaseName(m_path.data());
-    if (!db.open())
-        throw RuntimeError(db.lastError().text());
+    {
+        auto db = QSqlDatabase::addDatabase(u"QSQLITE"_qs, m_connectionName);
+        db.setDatabaseName(m_path.data());
+        if (!db.open())
+            throw RuntimeError(db.lastError().text());
+
+        int64_t transactedJobsCount = 0;
+        while (true)
+        {
+            m_jobsMutex.lock();
+            if (m_jobs.empty())
+            {
+                if (transactedJobsCount > 0)
+                {
+                    db.commit();
+                    m_dbLock.unlock();
+
+                    qDebug() << "Resume data changes are commited. Transacted jobs:" << transactedJobsCount;
+                    transactedJobsCount = 0;
+                }
+
+                if (isInterruptionRequested())
+                {
+                    m_jobsMutex.unlock();
+                    break;
+                }
+
+                m_waitCondition.wait(&m_jobsMutex);
+                if (isInterruptionRequested())
+                {
+                    m_jobsMutex.unlock();
+                    break;
+                }
+
+                m_dbLock.lockForWrite();
+                if (!db.transaction())
+                {
+                    LogMsg(tr("Couldn't begin transaction. Error: %1").arg(db.lastError().text()), Log::WARNING);
+                    m_dbLock.unlock();
+                    break;
+                }
+            }
+            std::unique_ptr<Job> job = std::move(m_jobs.front());
+            m_jobs.pop();
+            m_jobsMutex.unlock();
+
+            job->perform(db);
```

Comment:
I got a SIGSEGV here on qBt exit in current master 34802362adcc4ff78fa16318c0912f8b00c7acf7. `job` was null. Can't reproduce again.

Reply:
Not sure mine is related, I got silent crash when exiting qbt (master branch).
To reproduce:
1. download windows binary from either github actions or appveyor (both will crash)
2. create a portable folder "profile"
3. start qbt and exit
4. find the crash message in Windows event viewer (or within a debugger)

Reply:
@Chocobo1 I believe you are experiencing #18303, I can confirm silent crash with an appveyor build.

Reply:
Added AppVeyor stack traces to https://github.com/qbittorrent/qBittorrent/issues/18303#issuecomment-1556130416

Reply:
> Can't reproduce again.

It looks like it's about a race condition, so it's hard to reproduce. However, I believe #19005 should fix it.